File: SOLID/LSP/LSP Rules/Signature Rules/C++/MethodArgumentRule.cpp
#include<bits/stdc++.h>

using namespace std;


class Parent{
    public:
        virtual void method(string s){
            cout<<"METHOD EXECUTING USING PARENT OBJECT"<<endl;
    }
};

class Child : public Parent{
    public:
        void method(string s) override{
            cout<<"METHOD EXECUTING USING CHILD OBJECT"<<endl;
    }
};

class Client{
    private:
        Parent* object;

    public:
        Client(Parent* object){
            this->object = object;
        }

        void print(string msg){
            object->method(msg);
        }
};

int main(){
    Parent* parent = new Parent();
    Child* child = new Child();

    Client* client = new Client(child);
    client->print("hello");
}

File: SOLID/LSP/LSP Rules/Signature Rules/JAVA/MethodArgumentRule.java
// package LSP Rules.Signature Rules;

/**
 * 1. Method Argument Rule 
        ○ The overridden method in the subclass must accept the same argument types as 
        the parent, or wider (a “broader” type up the inheritance chain). 
        ○ Example: If the parent method takes a String, the child override must also take 
        String (or a supertype, e.g., Object), never an unrelated type like Integer. 
 */


        /*
            METHOD ARGUMENT RULE : 
            subtype argument types must be identical or broader
            java enforces this by ensuring identical method types/signatures during overrides. 
        */

class Parent{
    void method1(String object){
        System.out.println("PARENT METHOD RUNNING..........");
    }
}

class Child extends Parent{
    @Override
    void method1(String object){
        System.out.println("CHILD METHOD RUNNING..........");
    }

    // method overloading... the overriden method is used , this overloaded method is never used, it is used only if the client directly deals with the concrete child class

    // void method1(Object object){
    //     System.out.println("CHILD METHOD IS RUNNING..........");
    // }
}

/**
 *  CLIENT ------------------>  PARENT
 *         -
 *           -
 *             -
 *                -
 *                  -
 *                    - >CHILD     
 */
class Client{
    Parent object;

    Client(Parent object){
        this.object = object;
    }

    void print(String s){
        object.method1(s);
    }
}

public class MethodArgumentRule {
    public static void main(String[] args) {
        Parent parent = new Parent();

        Child child = new Child();

        Client client = new Client(child);
        client.print("hello");
        
    }
}
